pkg/logql/log/filter.go

```diff
@@ -90,6 +89,68 @@ func (a andFilter) ToStage() Stage {
 	}
 }
 
+type andFilters struct {
+	filters []Filterer
+}
+
+// NewAndFilters creates a new filter which matches only if all filters match
+func NewAndFilters(filters []Filterer) Filterer {
+	containsAllFilter := containsAllFilter{}
+	regexpFilters := make([]Filterer, 0)
+	n := 0
+	for _, filter := range filters {
+		// Make sure we take care of panics in case a nil or noop filter is passed.
+		if !(filter == nil || filter == TrueFilter) {
+			switch c := filter.(type) {
+			case containsFilter:
+				// Join all contain filters.
+				containsAllFilter.Add(c)
+			case regexpFilter:
+				regexpFilters = append(regexpFilters, c)
+			default:
+				// Keep filter
+				filters[n] = filter
+				n++
+			}
+		}
+	}
+	filters = filters[:n]
+
+	if !containsAllFilter.Empty() {
+		filters = append(filters, containsAllFilter)
```

Comment:
Seems like we're moving containsAllFilter at the end here ?

Reply:
Yes but before the regular expressions. Should I change this to
```suggestion
		filters = append([]Filterer{containsAllFilter}, filters...)
```

Reply:
The problem I see here is that If I write this:

`{app="foo"} != "timeout" |= "foo" |= "bar"` or `{app="foo"} |= "buzz" != "timeout" |= "foo" |= "bar"` we're shuffling execution order and that become a problem if I did it in a specific order because of performance.

While I think it's fine to push regexp at the end, I wonder if we could ensure basic filter order.

I think you need to unwrap the NotFilter and support that in the `containsAllFilter` WDYT ?

Reply:
aka I think we should only merge successive contain/not_contain only, and keep the order. If a IP filter comes first I wouldn't move it too.

Reply:
Ah, I think you are correct. We should not over optimize here.